gk/main.c

```diff
@@ -52,25 +435,118 @@ gk_proc(void *arg)
 		 */
 
 		/* Get burst of RX packets, from first port of pair. */
+		int i;
+		int ret = -1;
 		uint16_t num_rx;
-		uint16_t num_tx;
-		struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+		uint16_t num_tx = 0;
+		uint16_t num_tx_succ;
+		struct rte_mbuf *rx_bufs[GATEKEEPER_MAX_PKT_BURST];
+		struct rte_mbuf *tx_bufs[GATEKEEPER_MAX_PKT_BURST];
 
-		num_rx = rte_eth_rx_burst(port_in, 0, bufs,
+		/* Load a set of packets from the front NIC. */
+		num_rx = rte_eth_rx_burst(port_in, 0, rx_bufs,
 			GATEKEEPER_MAX_PKT_BURST);
 
 		if (unlikely(num_rx == 0))
 			continue;
 
+		for (i = 0; i < num_rx; i++) {
+			struct ipacket packet;
+			/*
+			 * Pointer to the flow entry in request state 
+			 * under evaluation.
+			 */
+			struct flow_entry *fe;
+			struct rte_mbuf *pkt = rx_bufs[i];
+			
+			extract_packet_info(pkt, &packet);
+
+			/* 
+			 * Find the flow entry for the IP pair.
+			 * Create a new flow entry if not found.
+			 */
+			ret = rte_hash_lookup_with_hash(instance->ip_flow_hash_table,
+				&packet.flow, pkt->hash.rss);
```

Comment:
The RSS initialization function ``gatekeeper_setup_rss()`` has been added to ``net.c``, this function can be called by both ``GK`` and ``GT``.